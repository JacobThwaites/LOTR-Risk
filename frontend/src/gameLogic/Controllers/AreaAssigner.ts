import { Player } from "../Models/Player";
import { AreaType } from "../Models/AreaType";

export class AreaAssigner {
  private players: Array<Player>;
  constructor(players: Array<Player>) {
    this.players = players;
  }

  assignAreas(areaLists: Array<AreaType[]>) {
    for (let i = 0; i < this.players.length; i++) {
      for (let j = 0; j < areaLists[i].length; j++) {
        this.addAreaToPlayer(areaLists[i][j], this.players[i]);
      }
      
    }
  }

  private addAreaToPlayer(area: AreaType, player: Player): void {
    player.addArea(area);
    area.setPlayer(player);
  }
}