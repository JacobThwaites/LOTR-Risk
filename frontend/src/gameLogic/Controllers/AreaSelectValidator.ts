import { Adjacencies } from "../Enums/AreaAdjacencies";
import { AreaType } from "../Models/AreaType";
import { Player } from "../Models/Player";

export function getAreasForUnitManeuvre(area: AreaType, player: Player): Array<AreaType> {
    const connected: Array<AreaType> = [];
    const visited: any = {};


    function dfs(area: AreaType) {
        const areaName = area.getName();
        if (visited[areaName]) {
            return;
        }

        visited[areaName] = true;

        if (!player.ownsArea(area)) {
            return;
        }

        connected.push(area);
        const adjacencies = Adjacencies[areaName];

        for (let i = 0; i < adjacencies.length; i++) {
            dfs(adjacencies[i]);
        }
    }

    dfs(area);
    
    return connected;
}