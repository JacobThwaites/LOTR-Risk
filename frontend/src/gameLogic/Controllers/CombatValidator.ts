import { Combat } from '../Enums/Combat'; 
import { AreaType } from '../Models/AreaType';
const { MAX_ATTACKING_DICE } = Combat;

export class CombatValidator {
    public static isCombatValid(attackingArea: AreaType, attackingDice: number): boolean {
        return this.validateAttackingDice(attackingArea, attackingDice);
    }

    private static validateAttackingDice(attackingArea: AreaType, attackingDice: number): boolean {
        const attackingAreaUnits = attackingArea.getUnits();
        return (
            attackingAreaUnits > attackingDice &&
            attackingDice <= MAX_ATTACKING_DICE &&
            attackingDice > 0
        );
    }
}