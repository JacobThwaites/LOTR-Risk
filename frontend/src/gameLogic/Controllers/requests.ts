export async function saveGame(numPlayers: number, playerAreas: Array<string>, playerName: string) {
    const players = formatPlayerData(playerAreas, playerName);

    try {
        const body = {
            numPlayers: numPlayers,
            players: players
        }

        return fetch('http://localhost:8000/api/game', {
            method: 'POST',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(body)
        });
    } catch (err) {
        console.error(err);
        return;
    }
}

function formatPlayerData(playerAreas: Array<string>, playerName: string): Array<{ name: string, areas: string }> {
    const players = [];

    const firstPlayer = {
        name: playerName,
        areas: playerAreas[0]
    }
    players.push(firstPlayer);

    for (let i = 1; i < playerAreas.length; i++) {
        const player = { name: '', areas: playerAreas[i] };
        players.push(player)
    }

    return players;
}

export async function getGame(gameID: string) {
    try {
        return fetch(`http://localhost:8000/api/game/${gameID}`, {
            method: 'GET',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            },
        });
    } catch (err) {
        console.error(err);
        return;
    }
}

export async function addUserIdToPlayer(playerID: string, userID: string) {
    try {
        const res = await fetch(`http://localhost:8000/api/player/${playerID}`, {
            method: 'PATCH',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({ userID })
        });

        if (!res.ok) {
            throw new Error("Faiedl to update player userID");
        }

        return true;

    } catch (err: any) {
        console.error(err.message);
        return false;
    }
}