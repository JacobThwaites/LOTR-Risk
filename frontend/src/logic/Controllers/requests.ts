export async function saveGame(numPlayers: number, playerAreas: Array<string | null>) {
  try {
    const body = {
        numPlayers: numPlayers,
        player1Areas: playerAreas[0],
        player2Areas: playerAreas[1],
        player3Areas: playerAreas[2],
        player4Areas: playerAreas[3]
    }

      return fetch('http://localhost:8000/api/game', {
          method: 'POST',
          headers: {
              'Accept': 'application/json',
              'Content-Type': 'application/json'
          },
          body: JSON.stringify(body)
      });
  } catch(err) {
      console.error(err);
      return;
  }
}

export async function getGame(gameUuid: string) {
    try {
        return fetch(`http://localhost:8000/api/game/${gameUuid}`, {
            method: 'GET',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            },
        });
    } catch(err) {
        console.error(err);
        return;
    }
  }