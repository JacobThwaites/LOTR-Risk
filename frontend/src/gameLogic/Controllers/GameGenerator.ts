import { Player } from '../Models/Player';
import { Game } from '../Models/Game'; 
import { AreaAssigner } from './AreaAssigner';
import { PlayerGenerator } from './PlayerGenerator';
import { AreaType } from '../Models/AreaType';

export class GameGenerator {
    public static generateGame(areaLists: Array<AreaType[]>, playerIDs: Array<number>): Game {
        const MAX_TURNS = 30;
        const players = this.generatePlayers(playerIDs);
        const areaAssigner = new AreaAssigner(players);
        areaAssigner.assignAreas(areaLists);
        const game = new Game(players, MAX_TURNS)
        game.assignStartingUnits();
        
        return game;
    }

    private static generatePlayers(playerIDs: Array<number>): Player[] {
        const playerGenerator = new PlayerGenerator(playerIDs);
        const players = playerGenerator.generatePlayers();
        return players;
    } 
}