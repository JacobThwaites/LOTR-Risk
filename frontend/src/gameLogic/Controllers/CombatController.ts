import { AreaType } from '../Models/AreaType';
import { CombatValidator } from './CombatValidator';

export class CombatController {
    private attackingArea: AreaType;
    private defendingArea: AreaType;
    constructor(attackingArea: AreaType, defendingArea: AreaType) {
        this.attackingArea = attackingArea;
        this.defendingArea = defendingArea;
    }

    getCombatResults(numAttackingDice: number, numDefendingDice: number) {
        const results = [];
        const attackingDice = this.rollDice(numAttackingDice);
        const defendingDice = this.rollDice(numDefendingDice);
        
        const firstResult = this.firstDiceCombat(attackingDice[0], defendingDice[0]);
        results.push(firstResult);

        if (numDefendingDice > 1) {
            const secondResult = this.secondDiceCombat(attackingDice[0], defendingDice[0]);
            results.push(secondResult);
        }

        return results;
    }

    firstDiceCombat(attackingDice: number, defendingDice: number) {
        const defenderBonus = this.getDefenderDiceBonus(this.defendingArea);
        const attackerBonus = this.getAttackerDiceBonus(this.attackingArea);
        // TODO: remove + 6
        const attackerScore = attackingDice + attackerBonus + 6;
        const defenderScore = defendingDice + defenderBonus - 6;
        if (attackerScore > defenderScore) {
            return 'attacker';
        } else {
            return 'defender';
        }
    }

    secondDiceCombat(attackerScore: number, defenderScore: number) {
        if (attackerScore > defenderScore) {
            return 'attacker';
        } else {
            return 'defender';
        }
    }

    handleResults(results: string[]) {
        for (let i = 0; i < results.length; i++) {
            if (results[i] === 'attacker') {
                this.removeUnitsFromLoser(this.defendingArea);
            } else {
                this.removeUnitsFromLoser(this.attackingArea);
            }  
        }

        this.checkDefendingUnitsRemaining();
    }

    removeUnitsFromLoser(losingArea: AreaType) {
        const losingPlayer = losingArea.getPlayer();
        losingPlayer!.removeUnits(1);
        losingArea.removeUnits(1);
    }

    isCombatValid(attackingDice: number, defendingDice: number): boolean {
        const combatValidator = new CombatValidator(this.attackingArea, this.defendingArea);
        return combatValidator.isCombatValid(attackingDice, defendingDice);
    } 

    rollDice(numberOfDice: number): Array<number> {
        let numbersRolled = [];
        for (let i = 0; i < numberOfDice; i++) {
            const number = this.getDiceRoll();
            numbersRolled.push(number); 
        }
        
        return numbersRolled.sort().reverse();
    }

    getDiceRoll(): number {
        const numberRolled = Math.floor(Math.random() * 6) + 1; 
        return numberRolled;
    }

    getDefenderDiceBonus(defendingArea: AreaType): number {
        let bonus = 0;
        bonus += defendingArea.getDefendingBonus();

        if (defendingArea.getHasLeader()) {
            bonus += 1;
        }
        return bonus;
    }

    getAttackerDiceBonus(attackingArea: AreaType): number {
        let bonus = 0;
        if (attackingArea.getHasLeader()) {
            bonus += 1;
        }
        return bonus;
    }

    checkDefendingUnitsRemaining() {
        const attacker = this.attackingArea.getPlayer();

        if (!attacker) {
            return;
        }

        if (this.defendingArea.getUnits() < 1) {
            attacker.addArea(this.defendingArea);
            this.defendingArea.setPlayer(attacker);
        }
    }
}