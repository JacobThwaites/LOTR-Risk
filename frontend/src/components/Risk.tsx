import React, { useEffect, useState } from "react";
import GameSetup from "./GameSetup";
import { Redirect } from "react-router";
import { saveGame } from "../logic/Controllers/requests";
import { convertPlayerAreasToString } from "../utils/playerAreaParser";
import { setupAreaAssignments } from "../logic/Controllers/AreaAssigner";

export default function Risk() {
    const [numberOfPlayers, setNumberOfPlayers] = useState(0);
    const [shouldDisplayChooseGameType, setShouldDisplayChooseGameType] = useState(true);
    const [shouldDisplayNumberOfPlayersSelector, setShouldDisplayNumberOfPlayersSelector] = useState(false);
    const [shouldDisplayGameSetup, setShouldDisplayGameSetup] = useState(true);
    const [playerName, setPlayerName] = useState('');
    const [gameType, setGameType] = useState('online');

    function onNumberSelect(number: number, b: any, a: { name: string }) {
        setNumberOfPlayers(number);
    }

    function onChangeName(event: any) {
        const playerName = event.target.value;
        setPlayerName(playerName);
    }

    function onSubmitNumberOfPlayers() {
        setShouldDisplayNumberOfPlayersSelector(false);
    }

    async function onSubmitName() {
        setShouldDisplayGameSetup(false);
    }

    function onSubmitGameType(gameType: string) {
        setGameType(gameType);
        setShouldDisplayChooseGameType(false);
        setShouldDisplayNumberOfPlayersSelector(true);
    }

    if (shouldDisplayGameSetup) {
        return (
            <GameSetup
                numberOfPlayers={numberOfPlayers}
                playerName={playerName}
                shouldDisplayChooseGameType={shouldDisplayChooseGameType}
                shouldDisplayNumberOfPlayersSelector={
                    shouldDisplayNumberOfPlayersSelector
                }
                onChangeNumberOfPlayers={onNumberSelect}
                onSubmitNumberOfPlayers={onSubmitNumberOfPlayers}
                onChangeName={onChangeName}
                onSubmitPlayerName={onSubmitName}
                onSubmitGameType={onSubmitGameType}
            />
        );
    }

    return <GameRedirect numberOfPlayers={numberOfPlayers} playerName={playerName} />
}

function GameRedirect(props: { numberOfPlayers: number, playerName: string }) {
    const [uuid, setUuid] = useState('');

    const getData = async () => {
        try {
            const areas = setupAreaAssignments(props.numberOfPlayers);
            const areaStrings = convertPlayerAreasToString(areas);
            const res = await saveGame(props.numberOfPlayers, areaStrings);
            const json = await res!.json()
            const { uuid } = json.data;
            setUuid(uuid);
        } catch (err) {
            console.error(err);
        }
    }

    useEffect(() => {
        getData();
    }, []);

    if (!uuid) {
        return <h1>Loading...</h1>
    }

    return (
        <Redirect
            to={{
                pathname: uuid,
                state: { numberOfPlayers: props.numberOfPlayers, playerName: props.playerName, gameUuid: uuid },
            }}
        />
    );
}