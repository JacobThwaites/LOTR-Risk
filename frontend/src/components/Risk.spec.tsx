import '@testing-library/jest-dom'
import React from 'react';
import { render, screen, fireEvent, act, waitFor } from '@testing-library/react'
import Risk from './Risk';
import { saveGame } from "../gameLogic/Controllers/requests";
import { MemoryRouter } from 'react-router-dom'; 

jest.mock('../gameLogic/Controllers/requests');

it('renders Risk correctly', async () => {
    const { container } = render(<Risk />);
    expect(container).toMatchSnapshot();
});

it('handles game setup', async () => {
    const mockGameID = 'asdf';
    const numberOfPlayers = 3;
    const mockResponse = {
        json: async () => ({ data: { uuid: mockGameID } }),
      };
      (saveGame as jest.Mock).mockResolvedValueOnce(mockResponse);

    jest.mock('react-router-dom', () => {
        return {
          Redirect: jest.fn(({ to }) => `Redirected to ${to}`),
        };
    });

    const { rerender, getByText } = render(
        <MemoryRouter>
            <Risk />
        </MemoryRouter>
    );
    
    const localGameTypeButton = await screen.findByTestId('game-type-selector--option_local');
    fireEvent.click(localGameTypeButton);
    
    const submitButton = await screen.findByTestId('game-type-selector--button');
    fireEvent.click(submitButton);
    
    rerender(
        <MemoryRouter>
            <Risk />
        </MemoryRouter>
    );
    
    const numberOfPlayersSelector = await screen.findByTestId('number-of-players--selector');
    fireEvent.change(numberOfPlayersSelector, { target: { value: numberOfPlayers} });
    
    const submitNumberOfPlayersButton = await screen.findByTestId('number-of-players--button');
    fireEvent.click(submitNumberOfPlayersButton);
    
    await act(async () => {
        // Verify that "Loading..." is displayed before the data is fetched
        expect(getByText('Loading...')).toBeInTheDocument();
    
        // Wait for the data to be fetched and state update to happen
        await waitFor(() => expect(saveGame).toHaveBeenCalledTimes(1));
    
        // Verify that the saveGame function was called with the correct argument
        expect(saveGame).toHaveBeenCalledWith(numberOfPlayers);

        // TODO: test that it redirects correctly
        
    })
});