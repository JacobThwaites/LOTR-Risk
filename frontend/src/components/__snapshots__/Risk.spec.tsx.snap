// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Risk correctly 1`] = `
<div>
  <div
    data-testid="game-type-selector"
    id="game-type-selector"
  >
    <h1
      id="game-type-selector--header"
    >
      Select Game Type
    </h1>
    <div
      id="game-type-selector--options"
    >
      <div
        class="game-type-selector--option"
      >
        <div
          class="radio-button-icon"
          id="game-type-selector--option_online"
        >
          <label
            class="radio-button-icon--container"
          >
            <input
              checked=""
              name="test"
              type="radio"
              value="online"
            />
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-globe fa-w-16 font-awesome-icon--checked"
              data-icon="globe"
              data-prefix="fas"
              data-testid="game-type-selector--option_online"
              focusable="false"
              role="img"
              viewBox="0 0 496 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M336.5 160C322 70.7 287.8 8 248 8s-74 62.7-88.5 152h177zM152 256c0 22.2 1.2 43.5 3.3 64h185.3c2.1-20.5 3.3-41.8 3.3-64s-1.2-43.5-3.3-64H155.3c-2.1 20.5-3.3 41.8-3.3 64zm324.7-96c-28.6-67.9-86.5-120.4-158-141.6 24.4 33.8 41.2 84.7 50 141.6h108zM177.2 18.4C105.8 39.6 47.8 92.1 19.3 160h108c8.7-56.9 25.5-107.8 49.9-141.6zM487.4 192H372.7c2.1 21 3.3 42.5 3.3 64s-1.2 43-3.3 64h114.6c5.5-20.5 8.6-41.8 8.6-64s-3.1-43.5-8.5-64zM120 256c0-21.5 1.2-43 3.3-64H8.6C3.2 212.5 0 233.8 0 256s3.2 43.5 8.6 64h114.6c-2-21-3.2-42.5-3.2-64zm39.5 96c14.5 89.3 48.7 152 88.5 152s74-62.7 88.5-152h-177zm159.3 141.6c71.4-21.2 129.4-73.7 158-141.6h-108c-8.8 56.9-25.6 107.8-50 141.6zM19.3 352c28.6 67.9 86.5 120.4 158 141.6-24.4-33.8-41.2-84.7-50-141.6h-108z"
                fill="currentColor"
              />
            </svg>
          </label>
        </div>
        <label>
          Online
        </label>
      </div>
      <div
        class="game-type-selector--option"
      >
        <div
          class="radio-button-icon"
          id="game-type-selector--option_local"
        >
          <label
            class="radio-button-icon--container"
          >
            <input
              name="test"
              type="radio"
              value="local"
            />
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-desktop fa-w-18 font-awesome-icon"
              data-icon="desktop"
              data-prefix="fas"
              data-testid="game-type-selector--option_local"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M528 0H48C21.5 0 0 21.5 0 48v320c0 26.5 21.5 48 48 48h192l-16 48h-72c-13.3 0-24 10.7-24 24s10.7 24 24 24h272c13.3 0 24-10.7 24-24s-10.7-24-24-24h-72l-16-48h192c26.5 0 48-21.5 48-48V48c0-26.5-21.5-48-48-48zm-16 352H64V64h448v288z"
                fill="currentColor"
              />
            </svg>
          </label>
        </div>
        <label>
          Local
        </label>
      </div>
    </div>
    <button
      class="custom-button"
      data-testid="game-type-selector--button"
      id="game-type-selector--button"
    >
      Submit
    </button>
  </div>
</div>
`;
